Add DDP tutorial series (video) (#2049)
pytorch/tutorials: beginner_source/ddp_fault_tolerance.rst

`Introduction <ddp_series_intro.html>`__ \|\| `What is DDP <ddp_theory.html>`__ \|\| `Single-node
Multi-GPU training <ddp_multigpu.html>`__ \|\| **Fault
Tolerance** \|\| `Multi-node

> training <../intermediate/ddp_multinode.html>`__ \|\| `mingpt training <../intermediate/ddp_minGPT.html>`__

Reviewer: In all files: 
Single-node Multi-GPU training => Single-Node Multi-GPU Training
Multi-node training => Multi-Node Training 
mingpt training => minGPT Training

      

View the code used in this video: https://github.com/pytorch/examples/blob/main/distributed/ddp-tutorial-series/multigpu_torchrun.py

.. raw:: html

   <div style="margin-top:10px; margin-bottom:10px;">
     <iframe width="560" height="315" src="https://www.youtube.com/embed/9kIvQOiwYzg" frameborder="0" allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
   </div>

Why use ``torchrun``
~~~~~~~~~~~~~~~~~~~~

``torchrun`` handles the minutiae of distributed training so that you
don't need to. For instance,

-  You don't need to set environment variables or explicitly pass the ``rank`` and ``world_size``; torchrun assigns this along with several other `environment variables <https://pytorch.org/docs/stable/elastic/run.html#environment-variables>`__.
-  No need to call ``mp.spawn`` in your script; you only need a generic ``main()`` entrypoint, and launch the script with ``torchrun``. This way the same script can be run in non-distributed as well as single-node and multinode setups. 
-  Gracefully restarting training from the last saved training snapshot

Graceful restarts
~~~~~~~~~~~~~~~~~~~~~

> For graceful restarts, it helps to have the following structure:

Reviewer: ```suggestion
For graceful restarts, you can use the following structure:
```

`Introduction <ddp_series_intro.html>`__ \|\| `What is DDP <ddp_theory.html>`__ \|\| `Single-node
Multi-GPU training <ddp_multigpu.html>`__ \|\| **Fault
Tolerance** \|\| `Multi-node
training <../intermediate/ddp_multinode.html>`__ \|\| `mingpt training <../intermediate/ddp_minGPT.html>`__

Fault-tolerant Distributed Training with ``torchrun``
=====================================================

Authors: `Suraj Subramanian <https://github.com/suraj813>`__

In distributed training, a single process failure can
disrupt the entire training job. Since the susceptibility for failure can be higher here, making your training
script robust is particularly important here. You might also prefer your training job to be *elastic* i.e. the ability to increase or decrease the number

> of processes in realtime while running the training job.

Reviewer 2: do we actually provide this realtime capability? i.e. if we set max nodes to 3 and min nodes to 1, and we start training with 1 node, and 1 more node comes alive, do we proactively pause training, re-rendezvous, and continue, or does there need to be some sort of failure that stops the training? 

cc @d4l3k

Author: From reading the docs, my sense is that all processes are terminated and restarted on all available nodes with re-rendezvous, to continue from the last available checkpoint. Is "realtime" inaccurate to describe this?

Author: rephrased to "ability to increase or decrease the number of processes during training."

PyTorch has a utility ``torchrun`` that provides fault-tolerance as well
as elastic training. When a failure occurs, torchrun logs the errors and
attempts to automatically restart all the processes from the last saved

> “snapshot” of the training job. 

Author: snapshot is not a variable or function used in code, so backticks are not appropriate here.

The snapshot saves more than just the model state; it can include
details about the number of epochs run, optimizer states or any other

> mutable attribute of the training job necessary for its continuity.

Reviewer 2: s/mutable/stateful maybe

.. grid:: 2

   .. grid-item-card:: :octicon:`mortar-board;1em;` What you will learn
      :shadow: none

      -  Launching multi-GPU training jobs with ``torchrun``
      -  Saving and loading snapshots of your training job
      -  Structuring your training script for graceful restarts

   .. grid-item-card:: :octicon:list-unordered;1em;` Prerequisites
      :shadow: none

      * High-level `overview <ddp_theory.html>`__ of DDP
      * Familiarity with `DDP code <ddp_multigpu.html>`__ 
      * A machine with multiple GPUs (this tutorial uses an AWS p3.8xlarge instance)
      * PyTorch `installed <https://pytorch.org/get-started/locally/>`__ with CUDA
      

View the code used in this video: https://github.com/suraj813/distributed-pytorch/blob/main/multigpu_torchrun.py

.. raw:: html

   <div style="margin-top:10px; margin-bottom:10px;">

>      <iframe width="560" height="315" src="https://www.youtube.com/embed/9kIvQOiwYzg" frameborder="0" allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Reviewer 3: Just curious, if this is hosted on PyTorch youtube channel or in ones personal file?

Author: On the PyTorch youtube channel

- No need to call ``mp.spawn`` in your script; you only need a generic
``main()`` entrypoint, and launch the script with ``torchrun``. This way
the same script can be run in non-distributed as well as single-node and
multinode setups. 
- Gracefully restarting training from the last saved training
snapshot

Graceful restarts
~~~~~~~~~~~~~~~~~~~~~
For graceful restarts, it helps to have the following structure:

.. code:: python

   def main():
     load_snapshot(snapshot_path)
     initialize()
     train()

   def train():
     for batch in iter(dataset):
       train_step(batch)

       if should_checkpoint:
         save_snapshot(snapshot_path)

If a failure occurs, torchrun will terminate all the processes and restart them. 
Each process entrypoint first loads and initializes the last saved snapshot, and continues training from there.
So at any failure, you only lose the training progress from the last saved snapshot. 

In elastic training, whenever there are any membership changes (adding or removing nodes), torchrun will terminate and spawn processes
on available devices. Having this structure ensures your training job can continue without manual intervention.

> Diff for `multigpu.py <https://github.com/suraj813/distributed-pytorch/blob/main/multigpu.py>`__ v/s `multigpu_torchrun.py <https://github.com/suraj813/distributed-pytorch/blob/main/multigpu_torchrun.py>`__

Reviewer 3: Can you please do not reference your personal git repositories, i.e. add `multigpu.py` either to the same PR or submit it to `[pytorch/examples](https://github.com/pytorch/examples)`

Author: I'll submit to the Examples repo
